File: services/graphql-crm/src/mutation/invoices.rs
use async_graphql::{Context, InputObject, Object};
use chrono::{DateTime, NaiveDate, Utc};
use rust_decimal::Decimal;
use uuid::Uuid;

use crate::models::enums::{InvoiceStatus, PaymentMethod};
use crate::models::invoices;

#[derive(Debug, Clone, InputObject)]
pub struct CreateInvoiceInput {
    pub opportunity_id: Option<Uuid>,
    pub status: Option<InvoiceStatus>,
    pub total: Option<Decimal>,
    pub issue_date: Option<NaiveDate>,
    pub due_date: Option<NaiveDate>,
    pub sent_at: Option<DateTime<Utc>>,
    pub paid_at: Option<DateTime<Utc>>,
    pub payment_method: Option<PaymentMethod>,
}

#[derive(Debug, Clone)]
pub struct Mutation;

#[Object(name = "CrmInvoicesMutations")]
impl Mutation {
    async fn create_invoice(
        &self,
        ctx: &Context<'_>,
        payload: CreateInvoiceInput,
    ) -> async_graphql::Result<invoices::Model> {
        todo!()
    }
    async fn update_invoice_opportunity_id(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        opportunity_id: Option<Uuid>,
    ) -> async_graphql::Result<invoices::Model> {
        todo!()
    }
    async fn update_invoice_status(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        status: Option<InvoiceStatus>,
    ) -> async_graphql::Result<invoices::Model> {
        todo!()
    }
    async fn update_invoice_total(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        total: Option<Decimal>,
    ) -> async_graphql::Result<invoices::Model> {
        todo!()
    }
    async fn update_invoice_issue_date(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        issue_date: Option<NaiveDate>,
    ) -> async_graphql::Result<invoices::Model> {
        todo!()
    }
    async fn update_invoice_due_date(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        due_date: Option<NaiveDate>,
    ) -> async_graphql::Result<invoices::Model> {
        todo!()
    }
    async fn update_invoice_sent_at(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        sent_at: Option<DateTime<Utc>>,
    ) -> async_graphql::Result<invoices::Model> {
        todo!()
    }
    async fn update_invoice_paid_at(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        paid_at: Option<DateTime<Utc>>,
    ) -> async_graphql::Result<invoices::Model> {
        todo!()
    }
    async fn update_invoice_payment_method(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        payment_method: Option<PaymentMethod>,
    ) -> async_graphql::Result<invoices::Model> {
        todo!()
    }
    async fn remove_invoice(&self, ctx: &Context<'_>, id: Uuid) -> async_graphql::Result<String> {
        todo!()
    }
}

File: src/graphql/mutations/crm/invoices.ts
import { graphql } from "@/lib/graphql/client";

export const createInvoice = graphql(`
  mutation CreateInvoice($payload: CreateInvoiceInput!) {
    crm {
      createInvoice(payload: $payload) {
        id
        dueDate
        issueDate
        paidAt
        sentAt
        status
        total
        paymentMethod
        createdAt
        updatedAt
        opportunityId
        opportunity {
          id
          name
        }
        items(limit: 10, page: 1) {
          id
          price
          quantity
          product {
            id
            name
          }
        }
      }
    }
  }
`);

export const updateInvoiceOpportunityId = graphql(`
  mutation UpdateInvoiceOpportunityId($id: UUID!, $opportunityId: UUID) {
    crm {
      updateInvoiceOpportunityId(id: $id, opportunityId: $opportunityId) {
        id
        dueDate
        issueDate
        paidAt
        sentAt
        status
        total
        paymentMethod
        createdAt
        updatedAt
        opportunityId
        opportunity {
          id
          name
        }
        items(limit: 10, page: 1) {
          id
          price
          quantity
          product {
            id
            name
          }
        }
      }
    }
  }
`);

export const updateInvoiceStatus = graphql(`
  mutation UpdateInvoiceStatus($id: UUID!, $status: InvoiceStatus) {
    crm {
      updateInvoiceStatus(id: $id, status: $status) {
        id
        dueDate
        issueDate
        paidAt
        sentAt
        status
        total
        paymentMethod
        createdAt
        updatedAt
        opportunityId
        opportunity {
          id
          name
        }
        items(limit: 10, page: 1) {
          id
          price
          quantity
          product {
            id
            name
          }
        }
      }
    }
  }
`);

export const updateInvoiceTotal = graphql(`
  mutation UpdateInvoiceTotal($id: UUID!, $total: Decimal) {
    crm {
      updateInvoiceTotal(id: $id, total: $total) {
        id
        dueDate
        issueDate
        paidAt
        sentAt
        status
        total
        paymentMethod
        createdAt
        updatedAt
        opportunityId
        opportunity {
          id
          name
        }
        items(limit: 10, page: 1) {
          id
          price
          quantity
          product {
            id
            name
          }
        }
      }
    }
  }
`);

export const updateInvoiceIssueDate = graphql(`
  mutation UpdateInvoiceIssueDate($id: UUID!, $issueDate: NaiveDate) {
    crm {
      updateInvoiceIssueDate(id: $id, issueDate: $issueDate) {
        id
        dueDate
        issueDate
        paidAt
        sentAt
        status
        total
        paymentMethod
        createdAt
        updatedAt
        opportunityId
        opportunity {
          id
          name
        }
        items(limit: 10, page: 1) {
          id
          price
          quantity
          product {
            id
            name
          }
        }
      }
    }
  }
`);

export const updateInvoiceDueDate = graphql(`
  mutation UpdateInvoiceDueDate($id: UUID!, $dueDate: NaiveDate) {
    crm {
      updateInvoiceDueDate(id: $id, dueDate: $dueDate) {
        id
        dueDate
        issueDate
        paidAt
        sentAt
        status
        total
        paymentMethod
        createdAt
        updatedAt
        opportunityId
        opportunity {
          id
          name
        }
        items(limit: 10, page: 1) {
          id
          price
          quantity
          product {
            id
            name
          }
        }
      }
    }
  }
`);

export const updateInvoiceSentAt = graphql(`
  mutation UpdateInvoiceSentAt($id: UUID!, $sentAt: DateTime) {
    crm {
      updateInvoiceSentAt(id: $id, sentAt: $sentAt) {
        id
        dueDate
        issueDate
        paidAt
        sentAt
        status
        total
        paymentMethod
        createdAt
        updatedAt
        opportunityId
        opportunity {
          id
          name
        }
        items(limit: 10, page: 1) {
          id
          price
          quantity
          product {
            id
            name
          }
        }
      }
    }
  }
`);

export const updateInvoicePaidAt = graphql(`
  mutation UpdateInvoicePaidAt($id: UUID!, $paidAt: DateTime) {
    crm {
      updateInvoicePaidAt(id: $id, paidAt: $paidAt) {
        id
        dueDate
        issueDate
        paidAt
        sentAt
        status
        total
        paymentMethod
        createdAt
        updatedAt
        opportunityId
        opportunity {
          id
          name
        }
        items(limit: 10, page: 1) {
          id
          price
          quantity
          product {
            id
            name
          }
        }
      }
    }
  }
`);

export const updateInvoicePaymentMethod = graphql(`
  mutation UpdateInvoicePaymentMethod($id: UUID!, $paymentMethod: PaymentMethod) {
    crm {
      updateInvoicePaymentMethod(id: $id, paymentMethod: $paymentMethod) {
        id
        dueDate
        issueDate
        paidAt
        sentAt
        status
        total
        paymentMethod
        createdAt
        updatedAt
        opportunityId
        opportunity {
          id
          name
        }
        items(limit: 10, page: 1) {
          id
          price
          quantity
          product {
            id
            name
          }
        }
      }
    }
  }
`);

export const addInvoiceItem = graphql(`
  mutation AddInvoiceItem($id: UUID!, $payload: CreateInvoiceItemInput!) {
    crm {
      addInvoiceItem(id: $id, payload: $payload) {
        id
        dueDate
        issueDate
        paidAt
        sentAt
        status
        total
        paymentMethod
        createdAt
        updatedAt
        opportunityId
        opportunity {
          id
          name
        }
        items(limit: 10, page: 1) {
          id
          price
          quantity
          product {
            id
            name
          }
        }
      }
    }
  }
`);

export const removeInvoiceItem = graphql(`
  mutation RemoveInvoiceItem($itemId: UUID!) {
    crm {
      removeInvoiceItem(itemId: $itemId) {
        id
        dueDate
        issueDate
        paidAt
        sentAt
        status
        total
        paymentMethod
        createdAt
        updatedAt
        opportunityId
        opportunity {
          id
          name
        }
        items(limit: 10, page: 1) {
          id
          price
          quantity
          product {
            id
            name
          }
        }
      }
    }
  }
`);

export const removeInvoice = graphql(`
  mutation RemoveInvoice($id: UUID!) {
    crm {
      removeInvoice(id: $id)
    }
  }
`);
